include/RAJA/exec-cuda/reduce_cuda.hxx

```diff
@@ -794,6 +794,430 @@ private:
+      }
+   }
+
+   //
+   // Operator to retrieve reduced min value (before object is destroyed).
+   // Accessor only operates on host.
+   //
+   operator T()
+   {
+      cudaErrchk(cudaDeviceSynchronize()) ;
+      m_reduced_val = static_cast<T>(m_blockdata[m_blockoffset].val);
+      return m_reduced_val;
+   }
+
+   //
+   // Operator to retrieve index value of min (before object is destroyed).
+   //
+   Index_type getMinLoc()
+   {
+      cudaErrchk(cudaDeviceSynchronize()) ; //it would be good not to call this
+      m_reduced_idx = m_blockdata[m_blockoffset].idx;
+      return m_reduced_idx;
+   }
+
+   //
+   // Min-loc function 
+   //
+   __device__
+   ReduceMinLoc< cuda_reduce<BLOCK_SIZE>, T> minloc(T val, Index_type idx) const 
+   {
+
+      __shared__ CudaReductionLocBlockDataType sd[BLOCK_SIZE];
+      __shared__ bool lastBlock;
+
+      // initialize shared memory
+      for ( int i = BLOCK_SIZE / 2; i > 0; i /=2 ) {     
+          // this descends all the way to 1
+          if ( threadIdx.x < i ) {                                
+             // no need for __syncthreads()
+             sd[threadIdx.x + i].val = m_reduced_val; 
+             sd[threadIdx.x + i].idx = m_reduced_idx; 
+          } 
+      }
+      __syncthreads();
+
+      sd[threadIdx.x].val = val;
+      sd[threadIdx.x].idx = idx; // need to reconcile loc vs idx naming
+      __syncthreads();
+
+      for (int i = BLOCK_SIZE / 2; i >= WARP_SIZE; i /= 2) {
+          if (threadIdx.x < i) {
+              sd[threadIdx.x] = RAJA_MINLOC(sd[threadIdx.x],sd[threadIdx.x + i]);
+          }
+          __syncthreads();
+      }
+
+      if (threadIdx.x < 16) {
+          sd[threadIdx.x] = RAJA_MINLOC(sd[threadIdx.x],sd[threadIdx.x+16]);
+      }
+      __syncthreads();
+
+      if (threadIdx.x < 8) {
+          sd[threadIdx.x] = RAJA_MINLOC(sd[threadIdx.x],sd[threadIdx.x+8]);
+      }
+      __syncthreads();
+
+      if (threadIdx.x < 4) {
+          sd[threadIdx.x] = RAJA_MINLOC(sd[threadIdx.x],sd[threadIdx.x+4]);
+      }
+      __syncthreads();
+
+      if (threadIdx.x < 2) {
+          sd[threadIdx.x] = RAJA_MINLOC(sd[threadIdx.x],sd[threadIdx.x+2]);
+      }
+      __syncthreads();
+
+      if (threadIdx.x < 1) {
+          lastBlock = false;
+          sd[threadIdx.x] = RAJA_MINLOC(sd[threadIdx.x],sd[threadIdx.x+1]);
+          m_blockdata[m_blockoffset + blockIdx.x+1]  = 
+              RAJA_MINLOC( sd[threadIdx.x], 
+                        m_blockdata[m_blockoffset + blockIdx.x+1] );
+          unsigned int oldBlockCount = atomicAdd(&retiredBlocks[m_myID],1);
+          lastBlock = (oldBlockCount == (gridDim.x-1));
+
+      }
+      __syncthreads();
+
+      if(lastBlock) {
+        if(threadIdx.x == 0) {
+          retiredBlocks[m_myID] = 0;
+        }
+
+        CudaReductionLocBlockDataType lmin;
+        lmin.val = m_reduced_val;
+        lmin.idx = m_reduced_idx;
+        for (int i = threadIdx.x; i < gridDim.x; i += BLOCK_SIZE) {
```

Comment:
Jeff on your comment I'm not sure what you're asking, gridDim.x here is the number of blocks retired, I could just as well name them slots as each block in the previous step updated its val,idx to it's own slot. Now for this loop there is only one block participating but in parallel each thread still participates modulo block size until it reaches gridDim.x